### PR TITLE
Lllnode mypy

### PR DESCRIPTION
**Pull request by jacqueswww**
### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/opcodes.py [hunk 1]
```diff
@@ -1,173 +1,174 @@
...
-    'CALLCODE': [0xf2, 7, 1, 700],
-    'RETURN': [0xf3, 2, 0, 0],
-    'DELEGATECALL': [0xf4, 6, 1, 700],
-    'CALLBLACKBOX': [0xf5, 7, 1, 700],
-    'SELFDESTRUCT': [0xff, 1, 0, 25000],
-    'STATICCALL': [0xfa, 6, 1, 40],
-    'REVERT': [0xfd, 2, 0, 0],
-    'INVALID': [0xfe, 0, 0, 0],
-    'DEBUG': [0xa5, 1, 0, 0]
+opcodes: Dict[str, Tuple[Union[int, None], int, int, int]] = {
```
Comment (charles-cooper) [inline vyper/opcodes.py]: I think the types of `opcodes` and `pseudo_opcodes` can be more specific, and then relax the type when we combine them into `comb_opcodes`.

File: vyper/optimizer.py [hunk 2]
```diff
@@ -28,11 +31,11 @@ def get_int_at(args, pos, signed=False):
...
 
 
-def int_at(args, pos):
-    return get_int_at(args, pos) is not None
+def int_at(args: List[LLLnode], pos: int, signed: bool = False) -> Optional[int]:
+    return get_int_at(args, pos, signed) is not None
 
 
-def search_for_set(node, var):
+def search_for_set(node: LLLnode, var: List[LLLnode]) -> bool:
```
Comment (charles-cooper) [inline vyper/optimizer.py]: I think this function is dead code anyways so we can delete it.

File: vyper/optimizer.py [hunk 3]
```diff
@@ -70,7 +75,7 @@ def optimize(node: LLLnode) -> LLLnode:
     argz = [optimize(arg) for arg in node.args]
     if node.value in arith and int_at(argz, 0) and int_at(argz, 1):
         left, right = get_int_at(argz, 0), get_int_at(argz, 1)
-        calcer, symb = arith[node.value]
+        calcer, symb = arith[str(node.value)]
```
Comment (charles-cooper) [inline vyper/optimizer.py]: can we leave a comment to the effect of
```
# `node.value in arith` implies that `node.value` is a `str`
```

File: vyper/optimizer.py [hunk 4]
```diff
@@ -93,7 +98,7 @@ def optimize(node: LLLnode) -> LLLnode:
             valency=node.valency,
         )
     elif _is_constant_add(node, argz):
-        calcer, symb = arith[node.value]
+        calcer, symb = arith[str(node.value)]
```
Comment (charles-cooper) [inline vyper/optimizer.py]: can we leave a comment to the effect of
```
# `_is_constant_add` implies that `node.value` is a `str`
```

File: vyper/optimizer.py [hunk 5]
```diff
@@ -139,14 +144,14 @@ def optimize(node: LLLnode) -> LLLnode:
             valency=node.valency,
         )
     elif node.value == "clamp" and int_at(argz, 0) and int_at(argz, 1) and int_at(argz, 2):
-        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):
+        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):  # type: ignore
```
Comment (charles-cooper) [inline vyper/optimizer.py]: why ignore the type?

File: vyper/optimizer.py [hunk 6]
```diff
@@ -139,14 +144,14 @@ def optimize(node: LLLnode) -> LLLnode:
...
             raise Exception("Clamp always fails")
-        elif get_int_at(argz, 1, True) > get_int_at(argz, 2, True):
+        elif get_int_at(argz, 1, True) > get_int_at(argz, 2, True):  # type: ignore
```
Comment (charles-cooper) [inline vyper/optimizer.py]: why ignore the type?

File: vyper/optimizer.py [hunk 7]
```diff
@@ -139,14 +144,14 @@ def optimize(node: LLLnode) -> LLLnode:
...
             raise Exception("Clamp always fails")
         else:
             return argz[1]
     elif node.value == "clamp" and int_at(argz, 0) and int_at(argz, 1):
-        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):
+        if get_int_at(argz, 0, True) > get_int_at(argz, 1, True):  # type: ignore
```
Comment (charles-cooper) [inline vyper/optimizer.py]: why ignore the type?

File: vyper/optimizer.py [hunk 8]
```diff
@@ -188,9 +193,9 @@ def optimize(node: LLLnode) -> LLLnode:
     # [ne, x, y] has the same truthyness as [xor, x, y]
     # rewrite 'ne' as 'xor' in places where truthy is accepted.
     elif has_cond_arg(node) and argz[0].value == 'ne':
-        argz[0] = LLLnode.from_list(['xor'] + argz[0].args)
+        argz[0] = LLLnode.from_list(['xor'] + argz[0].args)  # type: ignore
```
Comment (charles-cooper) [inline vyper/optimizer.py]: why ignore the type?

File: vyper/parser/lll_node.py [hunk 9]
```diff
@@ -183,10 +195,17 @@ def __init__(self,
                         f"be zerovalent: {self.args[3]}"
                     ))
                 self.valency = 0
+                rounds: int
                 if self.args[1].value in ('calldataload', 'mload') or self.args[1].value == 'sload':
-                    rounds = self.args[2].value
+                    if isinstance(self.args[2].value, int):
+                        rounds = self.args[2].value
+                    else:
+                        raise CompilerPanic('Unsupported rounds argument type.')
```
Comment (charles-cooper) [inline vyper/parser/lll_node.py]: might help to include `self.args[2]` in the error message

File: vyper/parser/lll_node.py [hunk 10]
```diff
@@ -183,10 +195,17 @@ def __init__(self,
...
                 else:
-                    rounds = abs(self.args[2].value - self.args[1].value)
+                    if isinstance(self.args[2].value, int) and isinstance(self.args[1].value, int):
+                        rounds = abs(self.args[2].value - self.args[1].value)
+                    else:
+                        raise CompilerPanic('Unsupported second argument types.')
```
Comment (charles-cooper) [inline vyper/parser/lll_node.py]: might help to include `self.args` in the error message

File: vyper/optimizer.py (hunk 5 above)
Comment (jacqueswww) [reply vyper/optimizer.py]: mypy is seriously dumb here; using the suggested hacks like above made the code look really messy.

File: vyper/optimizer.py (hunk 6 above)
Comment (jacqueswww) [reply vyper/optimizer.py]: same as above.

**Comment by jacqueswww** [comment]
@charles-cooper @fubuloubu Please check now; I rebased and fixed above. Also had to make args part of the CompilerPanic msg because it's not a ParserException.

**Comment by fubuloubu** [comment]
@jacqueswww looks like you need to rebase again?

**Pull request merged**